### PR TITLE
created user login endpoint and associate each user with its own posts

### PR DESCRIPTION
**Pull request by sangkips**
# Description

Please include a summary of the changes and the related issue. Please also include relevant motivation and context. List any dependencies that are required for this change.

I created a user login endpoint and associate each logged-in user with the posts created. You have to be the owner in order to do update and delete functionality.

Fixes # (issue)

## Type of change

- [ ] New feature (non-breaking change which adds functionality)

## Checklist:

- [x] My code follows the style guidelines of this project
- [x] I have performed a self-review of my code
- [x] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [x] My changes generate no new warnings
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] New and existing unit tests pass locally with my changes
- [ ] Any dependent changes have been merged and published in downstream modules

### CODE REVIEW

File: accounts/permissions.py [hunk 1]
```diff
@@ -0,0 +1,14 @@
+from rest_framework import permissions
+
+
+class IsOwner(permissions.BasePermission):
+    """ 
+    custom permission to only allow owners of object to edit, delete or update
+    """
```
Comment (JimmyTron) [inline accounts/permissions.py:7]: clean approach

File: CMS/urls.py [hunk 2]
```diff
@@ -35,10 +35,9 @@
     path('api/token/', MyTokenObtainPairView.as_view(), name='token_obtain_pair'),
     path('api/token/refresh/', TokenRefreshView.as_view(), name='token_refresh'),
 
-    path('blog/',include('blog.urls')),
+    path('api/', include('blog.urls')),
 
-
-    path("", include("accounts.urls")),
+    path("api/", include("accounts.urls")),
```
Comment (JimmyTron) [inline CMS/urls.py:40]: mbona hivi?

File: accounts/urls.py [hunk 3]
```diff
@@ -5,10 +5,12 @@
 router = routers.DefaultRouter()
 router.register('users', views.UserViewSet, basename='users')
 router.register('register', views.RegisterAccountViewSet, basename='register')
+router.register('login', views.UserLoginViewSet, basename='login')
 
 """Nesting to enable an account to be created under an existing user"""
 users_router = routers.NestedDefaultRouter(router, 'users', lookup='user')
-users_router.register('add_account', views.AddUserAccountViewSet, basename='add_account')
+users_router.register(
+    'add_account', views.AddUserAccountViewSet, basename='add_account')
```
Comment (JimmyTron) [inline accounts/urls.py:13]: why the overflow

File: blog/models.py [hunk 4]
```diff
@@ -1,45 +1,55 @@
...
     slug = models.SlugField(max_length=100, unique=True)
     description = models.CharField(max_length=500, null=True, blank=True)
     content = RichTextUploadingField()
-    created_by = models.ForeignKey(Account, on_delete=models.SET_NULL, null=True)
+    owner = models.ForeignKey(
+        settings.AUTH_USER_MODEL, on_delete=models.CASCADE)
     created_at = models.DateTimeField(auto_now_add=True)
     updated_at = models.DateTimeField(auto_now=True)
-    status = models.IntegerField(choices=STATUS, default=0) #"""This here serves to indicate whether a stori has been published or not."""
-    category = models.ForeignKey(Category,on_delete=models.PROTECT)
```
Comment (JimmyTron) [inline blog/models.py:33]: you like them open not closed in

File: accounts/permissions.py [hunk 5]
```diff
@@ -0,0 +1,14 @@
...
+
+    def has_object_permission(self, request, view, obj):
+        # allow GET methods
+        if request.method in permissions.SAFE_METHODS:
+            return True
+        # PUT, DELETE only allowed to the owner
+        return obj.owner == request.user
```
Comment (JimmyTron) [inline accounts/permissions.py:14]: why not bind it to the user account instead?

File: blog/views.py [hunk 6]
```diff
@@ -1,15 +1,26 @@
...
+from rest_framework import permissions
+from rest_framework.generics import ListAPIView
+from rest_framework import viewsets
+
 from blog.serializers import BlogSerializer
 from blog.filters import StoriFilter
 from blog.models import Stori
-from rest_framework.throttling import UserRateThrottle
 from .throttles import BlogRateThrottle
+from accounts.permissions import IsOwner
```
Comment (JimmyTron) [inline blog/views.py:11]: following

File: blog/views.py [hunk 7]
```diff
@@ -1,15 +1,26 @@
...
+class StoriViewSet(viewsets.ModelViewSet):
+    queryset = Stori.objects.all()
     serializer_class = BlogSerializer
     throttle_classes = [UserRateThrottle, BlogRateThrottle]
-    filter_backends = (DjangoFilterBackend,)   
-    filterset_class = StoriFilter
\ No newline at end of file
+    filter_backends = (DjangoFilterBackend,)
+    filterset_class = StoriFilter
+    permission_classes = [permissions.IsAuthenticatedOrReadOnly, IsOwner]
```
Comment (JimmyTron) [inline blog/views.py:21]: nice, was looking forward to see it get called

File: blog/views.py [hunk 8]
```diff
@@ -1,15 +1,26 @@
...
+
+    # Associate user with the post
+
+    def perform_create(self, serializer):
+        serializer.save(owner=self.request.user)
```
Comment (JimmyTron) [inline blog/views.py:26]: let's use account if possible. we want to separate contexts

**Review by JimmyTron — CHANGES_REQUESTED** [review]
The implementation is a straight forward one but why use the user model instead of the accounts since we will be having numerous accounts instead of users

File: accounts/permissions.py (hunk 1 above)
Comment (sangkips) [reply accounts/permissions.py:7]: Accounts details belong to a particular user, so whenever a post is created who do you want to associate that post with? Is it the account or the user? Tell me how will you relate the account and the token? Think it through even without referring to the ERD we were given..

File: CMS/urls.py (hunk 2 above)
Comment (sangkips) [reply CMS/urls.py:40]: > mbona hivi?

Just creating a better appealing URL, like /api/users or /api/blog/

File: accounts/permissions.py (hunk 5 above)
Comment (sangkips) [reply accounts/permissions.py:14]: > why not bind it to the user account instead?

It falls back to the first question you asked, I am relating the owner of the post and the user model

File: accounts/permissions.py (hunk 1 above)
Comment (Fredrick-Waihenya) [reply accounts/permissions.py:7]: Won't that create an issue, say a user has 3 accounts how will you specify if the deletion is of a specific post in one particular account. The user should have the ability to create or delete an account, but only the account has access to the blogs, simple is better.

File: CMS/urls.py (hunk 2 above)
Comment (hellen-22) [reply CMS/urls.py:40]: What if there's like an empty url in both account.urls and blog.urls won't there be like a conflict in the urls?

File: accounts/permissions.py (hunk 1 above)
Comment (hellen-22) [reply accounts/permissions.py:7]: > Won't that create an issue, say a user has 3 accounts how will you specify if the deletion is of a specific post in one particular account. The user should have the ability to create or delete an account, but only the account has access to the blogs, simple is better.

I think the deletion can be specified by the corresponding foreign keys/primary keys.

File: accounts/permissions.py (hunk 5 above)
Comment (hellen-22) [reply accounts/permissions.py:14]: But a post should be related to an account...?? Or what's happening.

File: CMS/urls.py (hunk 2 above)
Comment (JimmyTron) [reply CMS/urls.py:40]: > What if there's like an empty url in both account.urls and blog.urls won't there be like a conflict in the urls?

So you also saw it..

**Comment by hellen-22** [comment]
While we still here, what methods should a login endpoint accept?

**Comment by sangkips** [comment]
> While we still here, what methods should a login endpoint accept?

Its a POST request

File: CMS/urls.py (hunk 2 above)
Comment (sangkips) [reply CMS/urls.py:40]: This is the small change I have made. 
```
  path('api/', include('blog.urls')),
  path("", include("accounts.urls")),
```
This is the outcome
Screenshot from 2023-02-10 09-05-36

File: accounts/permissions.py (hunk 5 above)
Comment (sangkips) [reply accounts/permissions.py:14]: What is the logic to be used here? How will we identify and associate the user writing the blog content? I am still having issues with the relationship between the user and the account. 
when you log in where do you go?

**Pull request closed without merging**